pkg/codegen/nodejs/gen.go

```diff
@@ -1866,32 +1984,136 @@ func getChildMod(modName string) string {
 	return child
 }
 
-// genIndex emits an index module, optionally re-exporting other members or submodules.
-func (mod *modContext) genIndex(exports []string) string {
-	w := &bytes.Buffer{}
-
-	var imports []string
-	// Include the SDK import if we'll be registering module resources.
-	if len(mod.resources) != 0 {
-		imports = mod.sdkImports(false /*nested*/, true /*utilities*/)
+func (mod *modContext) nodeModuleOptimizationFlags() struct {
+	lazyLoadFunctions          bool
+	lazyLoadResources          bool
+	useTypeOnlyEnumsReferences bool
+} {
+	flags := struct {
+		lazyLoadFunctions          bool
+		lazyLoadResources          bool
+		useTypeOnlyEnumsReferences bool
+	}{}
+	nodePackageInfo := NodePackageInfo{}
+	if languageInfo, ok := mod.pkg.Language["nodejs"]; ok {
+		if info, ok2 := languageInfo.(NodePackageInfo); ok2 {
+			nodePackageInfo = info
+		}
+	}
+
+	for _, s := range nodePackageInfo.OptimizeNodeModuleLoading {
+		switch s {
+		case "lazy-load-functions":
+			flags.lazyLoadFunctions = true
+		case "lazy-load-resources":
+			flags.lazyLoadResources = true
+		case "use-type-only-enums-references":
+			flags.useTypeOnlyEnumsReferences = true
+		default:
+			continue
+		}
 	}
-	mod.genHeader(w, imports, nil, nil)
+	return flags
+}
 
-	// Export anything flatly that is a direct export rather than sub-module.
-	if len(exports) > 0 {
-		modDir := strings.ToLower(mod.mod)
-		fmt.Fprintf(w, "// Export members:\n")
-		sort.Strings(exports)
-		for _, exp := range exports {
-			rel, err := filepath.Rel(modDir, exp)
-			contract.Assert(err == nil)
-			if path.Base(rel) == "." {
-				rel = path.Dir(rel)
-			}
-			fmt.Fprintf(w, "export * from \"./%s\";\n", strings.TrimSuffix(rel, ".ts"))
+func (mod *modContext) genReexport(w io.Writer, exp fileInfo) {
+	modDir := strings.ToLower(mod.mod)
+	rel, err := filepath.Rel(modDir, exp.pathToNodeModule)
+	contract.Assert(err == nil)
+	if path.Base(rel) == "." {
+		rel = path.Dir(rel)
+	}
+	quotedImport := fmt.Sprintf(`"./%s"`, strings.TrimSuffix(rel, ".ts"))
+	flags := mod.nodeModuleOptimizationFlags()
+	if exp.fileType == functionFileType && flags.lazyLoadFunctions {
+		mod.genFunctionReexport(w, exp.functionFileInfo, quotedImport)
+	} else if exp.fileType == resourceFileType {
+		if flags.lazyLoadResources {
+			// optimize lazy-loading resource files
+			mod.genResourceReexport(w, exp.resourceFileInfo, quotedImport)
+		} else {
+			fmt.Fprintf(w, "export * from %s;\n", quotedImport)
+
+			// Something relies on these to be imported also.
+			fmt.Fprintf(w, "import { %s } from %s;\n",
+				exp.resourceFileInfo.resourceClassName,
+				quotedImport)
 		}
+	} else {
+		// non-optimized but foolproof eager reexport
+		fmt.Fprintf(w, "export * from %s;\n", quotedImport)
+	}
+}
```

Comment:
This function is rather challenging to follow. Not sure what to do about that, maybe a few more comments? I'm unsure.